File: exercism/haskell/Robot.hs
{-# LANGUAGE TemplateHaskell #-}

module Robot
    ( Bearing(East,North,South,West)
    , bearing
    , coordinates
    , mkRobot
    , move
    ) where

import           Lens.Micro    (_1, _2, (%~), (&), (^.))
import           Lens.Micro.TH (makeLenses)

data Bearing = North
             | East
             | South
             | West
             deriving (Eq, Show)

data Robot = Robot { _co :: (Integer, Integer), _dir :: Bearing }

makeLenses ''Robot

coordinates :: Robot -> (Integer, Integer)
coordinates = (^. co)

bearing :: Robot -> Bearing
bearing = (^. dir)

mkRobot :: Bearing -> (Integer, Integer) -> Robot
mkRobot direction coo =
  Robot { _co = coo
        , _dir = direction
        }

turnRight :: Bearing ->  Bearing
turnRight North = East
turnRight East  = South
turnRight South = West
turnRight West  = North

turnLeft :: Bearing -> Bearing
turnLeft East  = North
turnLeft South = East
turnLeft West  = South
turnLeft North = West

advance :: Bearing -> (Integer, Integer) -> (Integer, Integer)
advance North = _2 %~ (+1)
advance East  = _1 %~ (+1)
advance South = _2 %~ subtract 1
advance West  = _1 %~ subtract 1

move :: Robot -> String -> Robot
move = foldl (flip mkMove)
  where
    mkMove :: Char -> Robot -> Robot
    mkMove 'R' r = r & dir %~ turnRight
    mkMove 'L' r = r & dir %~ turnLeft
    mkMove 'A' r = r & co %~ advance (r^.dir)
    mkMove _ r   = r

File: exercism/rust/robot.rs
// The code below is a stub. Just enough to satisfy the compiler.
// In order to pass the tests you can add-to or change any of this code.

#[derive(PartialEq, Debug)]
pub enum Direction {
    North,
    East,
    South,
    West,
}

pub struct Robot {
    x: i32,
    y: i32,
    d: Direction,
}

impl Robot {
    pub fn new(x: i32, y: i32, d: Direction) -> Self {
        Robot { x, y, d }
    }

    #[must_use]
    pub fn turn_right(self) -> Self {
        match self.d {
            Direction::North => Self { d: Direction::East, ..self },
            Direction::East => Self { d: Direction::South, ..self },
            Direction::South => Self { d: Direction::West, ..self },
            Direction::West => Self { d: Direction::North, ..self },
        }
    }

    #[must_use]
    pub fn turn_left(self) -> Self {
        match self.d {
            Direction::North => Self { d: Direction::West, ..self },
            Direction::East => Self { d: Direction::North, ..self },
            Direction::South => Self { d: Direction::East, ..self },
            Direction::West => Self { d: Direction::South, ..self },
        }
    }

    #[must_use]
    pub fn advance(self) -> Self {
        match self.d {
            Direction::North => Self { y: self.y + 1, ..self },
            Direction::East => Self { x: self.x + 1, ..self },
            Direction::South => Self { y: self.y - 1, ..self },
            Direction::West => Self { x: self.x - 1, ..self },
        }
    }

    #[must_use]
    pub fn instructions(self, instructions: &str) -> Self {
        instructions.chars().fold(self, |acc, c| acc.step(c))
    }

    pub fn step(self, c: char) -> Self {
        match c {
            'R' => self.turn_right(),
            'A' => self.advance(),
            'L' => self.turn_left(),
            _ => unreachable!()
        }
    }

    pub fn position(&self) -> (i32, i32) {
        (self.x, self.y)
    }

    pub fn direction(&self) -> &Direction {
        &self.d
    }
}
